Python script automate tf-m build build process

urutva opened the pull request:
### Description 
#### Summary of change <!-- Required -->
The python script (tools/psa/build_tfm.py) automates tf-m build process
for a particular target or all supported tf-m targets and copies generated
binary to a predefined location.

Updated documentation describing steps to add single v7-m  and dual v7-M
targets and building tf-m using python script.

Added following attributes to dual v7-M and v8-M SPE generic target,

* inherits: PSA generic target PSA_DUAL_V7_M_SPE
* tfm_target_name: Target name in TF-M
* tfm_bootloader_supported: If TF-M bootloader is supported by the target.
			    Values supported are "true" and "false"
* tfm_default_toolchain: Default TF-M toolchain supported. Values supported
			 are "ARMCLANG" and "GNUARM"
* tfm_supported_toolchains: Supported TF-M toolchains. Values supported
			    are "ARMCLANG" and "GNUARM"
* delivery_dir: The directory to which TF-M binary will be copied to

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

#### Documentation <!-- Optional, but most likely you need it -->

<!-- Details of any document updates required, including links to PR against the docs repository -->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Required
    Provide all the information required, listing all the testing performed. For new targets please attach full test results
    for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    Manually tested. Another PR will be raised with pytest test cases for tf-m build tool
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Optional
    Request additional reviewers with @username or @team
-->
@Patater @jainvikas8 
----------------------------------------------------------------------------------------------------------------
### Release Notes <!-- Required for features, deprecations, breaking changes and other major PRs -->

<!--
    All 3 sections are compulsory for Major PR types. For Feature PRs only the summary section is required.
    This section is automatically added to release notes. Please fill in each sub-section with sufficient detail for a user.
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types). 
-->

#### Summary of changes
Python script automate tf-m build build process

#### Impact of changes
* The mbed-os release process should be updated to run tools/psa/build_tfm.py to generate tf-m binary for supported targets.
* New dual v7-M and v8-M targets include following new attributes,
    * inherits: PSA generic target PSA_DUAL_V7_M_SPE
    * tfm_target_name: Target name in TF-M
    * tfm_bootloader_supported: If TF-M bootloader is supported by the target.
			                             Values supported are "true" and "false"
    * tfm_default_toolchain: Default TF-M toolchain supported. Values supported
			                    are "ARMCLANG" and "GNUARM"
    * tfm_supported_toolchains: Supported TF-M toolchains. Values supported
			                            are "ARMCLANG" and "GNUARM"
    * delivery_dir: The directory to which TF-M binary will be copied to
* Exporters are not supported for TF-M targets

#### Migration actions required
None

urutva committed:
Python script automate tf-m build build process

The python script (tools/psa/build_tfm.py) automates tf-m build process
for a particular target or all supported tf-m targets and copies generated
binary to a predefined location.

Updated documentation describing steps to add single v7-m  and dual v7-M
targets and building tf-m using python script.

Added following attributes to dual v7-M and v8-M SPE generic target,

* inherits: PSA generic target PSA_DUAL_V7_M_SPE
* tfm_target_name: Target name in TF-M
* tfm_bootloader_supported: If TF-M bootloader is supported by the target.
			    Values supported are "true" and "false"
* tfm_default_toolchain: Default TF-M toolchain supported. Values supported
			 are "ARMCLANG" and "GNUARM"
* tfm_supported_toolchains: Supported TF-M toolchains. Values supported
			    are "ARMCLANG" and "GNUARM"
* delivery_dir: The directory to which TF-M binary will be copied to

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
Validating Mbed OS target default attributes

Do not check for Mbed OS target default attributes in case of targets
which supports TF-M

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -1,15 +1,19 @@
 
-## TODO
-* Adding dual v7-M targets
+*Note*  
```
Glossary, not Note

urutva replied:
@Patater Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -1,15 +1,19 @@
 # Supporting PSA in Mbed OS
 
-This document is still a work in progress. It describes the process of adding new PSA targets to Mbed OS. It also outlines the new directory structure created to support PSA on single v7-M, dual v7-M and single v8-M.
+This document is still a work in progress. It describes the process of adding new PSA targets to Mbed OS and building [TF-M](https://git.trustedfirmware.org/trusted-firmware-m.git/tree/?h=feature-twincpu) for dual v7-M and single v8-M SPE. It also outlines the new directory structure created to support PSA on single v7-M, dual v7-M and single v8-M.
```
```diff
- for dual v7-M and single v8-M SPE
+ for v8-M and twin CPU targets.
```

Patater replied:
```diff
- support PSA on single v7-M, dual v7-M and single v8-M.
+ support PSA on all targets.
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -1,15 +1,19 @@
-## Adding new PSA targets
-To help with the creation of new PSA targets, couple of generic PSA targets have been added to `targets/targets.json`.
+# Adding new PSA targets
+To help with the creation of new PSA targets, couple of generic targets have been added to `targets/targets.json`.
```
```diff
- To help with the creation of new PSA targets, couple of
+ To help with the creation of new PSA targets, a couple of
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -18,16 +22,32 @@ To help with the creation of new PSA targets, couple of generic PSA targets have
 * `PSA_V8_M_SPE` (v8-M SPE generic target)
 
-The new PSA targets **MUST** `inherit` from one of these targets depending on the architecture:
+A Single v7-M target doesn't support hardware separation to isolate NSPE from SPE. Therefore for single v7-M targets only NSPE target is defined. The PSA API compliance is achieved with the help of emulation of PSA secure services.
```
```diff
- A Single v7-M target doesn't support hardware separation to isolate NSPE from SPE.
+ A single-core PSA target doesn't employ hardware isolation between the NSPE and the SPE.
```

Patater replied:
```diff
- The PSA API compliance is achieved with the help of emulation of PSA secure services.
+ PSA secure service emulation enables PSA API compatibility.
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -18,16 +22,32 @@ To help with the creation of new PSA targets, couple of generic PSA targets have
-The new PSA targets **MUST** `inherit` from one of these targets depending on the architecture:
+A Single v7-M target doesn't support hardware separation to isolate NSPE from SPE. Therefore for single v7-M targets only NSPE target is defined. The PSA API compliance is achieved with the help of emulation of PSA secure services.
+
+A Dual v7-M target will have at least two Armv7-M cores to isolate NSPE from SPE. On dual v7-M targets, TF-M will be running on SPE providing PSA services. Therefore, it is **MANDATORY** to define both SPE and NSPE target for dual v7-M targets.
```
```diff
- A Dual v7-M target will have at least two Armv7-M cores to isolate NSPE from SPE.
+ A dual-core PSA target will have at least two cores that are either Armv7-M or Armv6-M. One core will be used for the SPE and another for the NSPE. Hardware isolation between the cores enables PSA compliance.
```

Patater replied:
```diff
- On dual v7-M targets, TF-M will be running on SPE providing PSA services.
+ On dual-core targets, TF-M runs on the SPE and provides PSA services.
```

Patater replied:
```diff
- define both SPE and NSPE target for dual v7-M targets.
+ define both the SPE and the NSPE target for dual-core PSA targets.
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -18,16 +22,32 @@ To help with the creation of new PSA targets, couple of generic PSA targets have
+
+A Dual v7-M target will have at least two Armv7-M cores to isolate NSPE from SPE. On dual v7-M targets, TF-M will be running on SPE providing PSA services. Therefore, it is **MANDATORY** to define both SPE and NSPE target for dual v7-M targets.
+
+A Armv8-M target supports hardware separation to isolate NSPE from SPE. On v8-M targets, TF-M will be running on SPE providing PSA services. Therefore, it is **MANDATORY** to define both SPE and NSPE target for v8-M targets.
```
```diff
- A Armv8-M target supports hardware separation to isolate NSPE from SPE.
+ An Armv8-M PSA target employs hardware to isolate the NSPE from the SPE.
```

Patater replied:
```diff
- On v8-M targets, TF-M will be running on SPE providing PSA services.
+ On Armv8-M targets, TF-M runs on the SPE and provides PSA services.
```

Patater replied:
```diff
- define both SPE and NSPE target for v8-M targets.
+ define both the SPE and the NSPE target for Armv8-M PSA targets.
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -18,16 +22,32 @@ To help with the creation of new PSA targets, couple of generic PSA targets have
+
+A Armv8-M target supports hardware separation to isolate NSPE from SPE. On v8-M targets, TF-M will be running on SPE providing PSA services. Therefore, it is **MANDATORY** to define both SPE and NSPE target for v8-M targets.
+
+The new PSA targets **MUST** `inherit` from one of the generic PSA targets depending on the architecture. Also, in case of dual v7-M and v8-M targets, SPE and NSPE targets **MUST** inherit from the respective SPE and NSPE PSA generic targets.
```
```diff
- from one of the generic PSA targets depending on the architecture.
+ from the generic PSA target that corresponds to that target's architecture.
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -18,16 +22,32 @@ To help with the creation of new PSA targets, couple of generic PSA targets have
+The new PSA targets **MUST** `inherit` from one of the generic PSA targets depending on the architecture. Also, in case of dual v7-M and v8-M targets, SPE and NSPE targets **MUST** inherit from the respective SPE and NSPE PSA generic targets.
 
-Example: 
+Example single v7-M: 
```
```diff
- single v7-M
+ single-core PSA target
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
````diff
@@ -18,16 +22,32 @@ To help with the creation of new PSA targets, couple of generic PSA targets have
 ```
 
-In addition, the new PSA target **MUST** `inherit` from `Target` which defines default target config options. The only exception is, if the new PSA target is inheriting from another generic target which inherits from `Target`.
+Example dual v7-M:
````
```diff
- dual v7-M
+ dual-core PSA target
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -40,8 +60,8 @@ Example:
-### Adding single v7-M targets
-The PSA on single v7-M is supported with the help of Mbed OS PSA services. The following is taken from `targets/targets.json` and shows a PSA enabled target, `K64F`.
+## Adding single v7-M targets
+A target can ge categorized as a single v7-M target if it only has a single Armv7-M core. The PSA on single v7-M is supported with the help of emulated PSA services implemented in Mbed OS. The following is taken from `targets/targets.json` and shows a PSA enabled single v7-M target, `K64F`.
```
```diff
- A target can ge categorized as a single v7-M target if it only has a single Armv7-M core.
+ A target can be categorized as a single-core target if it only has a single core.
```

Patater replied:
```diff
- The PSA on single v7-M is supported with the help of emulated PSA services implemented in Mbed OS.
+ Mbed OS's PSA service emulation provides PSA compatibility for single-core PSA targets.
```

Patater replied:
```diff
- shows a PSA enabled single v7-M target
+ shows a PSA-enabled single-core target
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -40,8 +60,8 @@ Example:
 
-### Adding single v7-M targets
-The PSA on single v7-M is supported with the help of Mbed OS PSA services. The following is taken from `targets/targets.json` and shows a PSA enabled target, `K64F`.
+## Adding single v7-M targets
```
```diff
- Adding single v7-M targets
+ Adding single-core PSA targets
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -104,11 +124,54 @@ The PSA on single v7-M is supported with the help of Mbed OS PSA services. The f
 Check [extra_labels](https://os.mbed.com/docs/mbed-os/v5.14/reference/adding-and-configuring-targets.html), [device_has](https://os.mbed.com/docs/mbed-os/v5.14/reference/adding-and-configuring-targets.html) and [features](https://os.mbed.com/docs/mbed-os/v5.14/reference/adding-and-configuring-targets.html) for more information.
 
-### Adding dual v7-M targets
+## Adding dual v7-M targets
```
```diff
- Adding dual v7-M targets
+ Adding dual-core PSA targets
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -104,11 +124,54 @@ The PSA on single v7-M is supported with the help of Mbed OS PSA services. The f
 
-### Adding dual v7-M targets
+## Adding dual v7-M targets
+A target can ge categorized as a dual v7-M target if it has at least two Armv7-M cores. On dual v7-M targets, TF-M will be running on SPE providing PSA services. Therefore, it is **MANDATORY** to define both SPE and NSPE target for dual v7-M targets.
```
```diff
- A target can ge categorized as a dual v7-M target if it has at least two Armv7-M cores.
+ A target can be categorized as a dual-core target if it has at least two cores that are either Armv7-M or Armv6-M.
```

Patater replied:
```diff
- On dual v7-M targets, TF-M will be running on SPE providing PSA services.
+ On dual-core PSA targets, TF-M runs on the SPE and provides PSA services.
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -104,11 +124,54 @@ The PSA on single v7-M is supported with the help of Mbed OS PSA services. The f
+## Adding dual v7-M targets
+A target can ge categorized as a dual v7-M target if it has at least two Armv7-M cores. On dual v7-M targets, TF-M will be running on SPE providing PSA services. Therefore, it is **MANDATORY** to define both SPE and NSPE target for dual v7-M targets.
+
+The SPE target **MUST** contain following attributes,
```
```diff
- contain following attributes,
+ contain the following attributes:
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -104,11 +124,54 @@ The PSA on single v7-M is supported with the help of Mbed OS PSA services. The f
+* inherits: PSA generic target PSA_DUAL_V7_M_SPE
+* tfm_target_name: Target name in TF-M
+* tfm_bootloader_supported: If TF-M bootloader is supported by the target. Values supported are "true" and "false"
+* tfm_default_toolchain: Default TF-M toolchain supported. Values supported are "ARMCLANG" and "GNUARM"
```
Do we need to invent new names like `ARMCLANG` and `GNUARM`? Can we re-use the existing toolchain names `ARMC` and `GCC_ARM`?

urutva replied:
@Patater The toolchain names `ARMCLANG` and `GNUARM` are used by TF-M. Therefore new Mbed OS targets that supports TF-M should have these names.

Other option is to continue using Mbed OS toolchain names and use key-value pair in the python script to identify TF-M toolchain names. Since the toolchains are target specific configuration I'd prefer define them in targets.json.

Patater replied:
OK, I am happy to use TF-M names then. The consumer is the TF-M build system. Best to match that.

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
````diff
@@ -104,11 +124,54 @@ The PSA on single v7-M is supported with the help of Mbed OS PSA services. The f
+    },
+```
+
+Please pay attention to config options `extra_labels_add` and `device_has_remove`. If needed then a PSA target definition **MUST** use [extra_labels/device_has]`_add` or [extra_labels/device_has]`_remove` (not `extra_labels` or `device_has`) to add/remove either extra_labels or target capabilities. Also, use `[feature_]`add/remove to add/remove a feature.
````
```diff
- Please pay attention to config options
+ Please pay attention to the config options
```

Patater replied:
```diff
- If needed then a PSA
+ If needed, a PSA
```

Patater replied:
Is "`[feature_]`add/remove" supposed to be "`feature_`[add/remove]"?

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -104,11 +124,54 @@ The PSA on single v7-M is supported with the help of Mbed OS PSA services. The f
+Check [extra_labels](https://os.mbed.com/docs/mbed-os/v5.14/reference/adding-and-configuring-targets.html), [device_has](https://os.mbed.com/docs/mbed-os/v5.14/reference/adding-and-configuring-targets.html) and [features](https://os.mbed.com/docs/mbed-os/v5.14/reference/adding-and-configuring-targets.html) for more information.
 
-### Adding v8-M targets
+## Adding v8-M targets
```
```diff
- Adding v8-M targets
+ Adding Armv8-M PSA targets
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
````diff
@@ -120,8 +183,36 @@ example mbed_app.json:
 ```
 
+# Building TF-M
+On dual v7-M and v8-M targets, TF-M will be running on SPE providing PSA services. A python script `tools/psa/build_tfm.py` automates building TF-M and copying the binary to predefined location defined by attribute `delivery_dir`.
````
```diff
- On dual v7-M and v8-M targets, TF-M will be running on SPE providing PSA services.
+ For dual-core and Armv8-M PSA targets, TF-M runs on the SPE and provides PSA services.
```

Patater replied:
```diff
- copying the binary to predefined location
+ copying the binary to a predefined location
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -120,8 +183,36 @@ example mbed_app.json:
+  -m {CY8CPROTO_064_SB_S}, --mcu {CY8CPROTO_064_SB_S}
+                        Build for the given MCU
+  -t {ARMCLANG,GNUARM}, --toolchain {ARMCLANG,GNUARM}
+                        Build for the given tool chain (default is
```
No space in "tool chain" please, here and in the Python script.

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -120,8 +183,36 @@ example mbed_app.json:
+  -l, --list            Print supported TF-M secure targets
+  --clone               Clone TF-M git repo and its dependencies
+  --commit              Create a git commit for each platform
+  --clean-build         Remove cmake build directory
```
For all uses of `cmake` that are supposed to be for humans:

```diff
- cmake
+ CMake
```

urutva replied:
Fixed

Patater replied:
```diff
- Remove cmake build directory
+ Remove CMake build directory
```

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -120,8 +183,36 @@ example mbed_app.json:
+
+If the python script is invoked without any options then TF-M will be built for all the supported targets and the binary will be copied to predefined location defined by attribute `delivery_dir`.
+
+# New directory structure
```
```diff
- New directory structure
+ PSA target directory structure
```

urutva replied:
Fixed

Patater commented on tools/psa/build_tfm.py:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
```
Can we not be compatible with both python2 and python3?

urutva replied:
@Patater With python2 going out of support in Jan2020, I'd prefer not to keep on supporting python2. This is one way of saying script is only compatible with python3.

Patater replied:
This may prevent us from merging into Mbed OS `master`, which still supports `python2`. I'm not sure when `python2` will be dropped from Mbed OS `master`.

Patater commented on tools/psa/build_tfm.py:
```diff
@@ -160,22 +181,301 @@ def clean_up_cloned_repos():
+    logger.debug("Found the following TF-M targets: {}".format(
+                                                ', '.join(tfm_secure_targets)))
+
+    return [get_target_info(t) for t in tfm_secure_targets]
```
Probably best to return a generator, not a list.

urutva replied:
Fixed

Patater commented on tools/psa/build_tfm.py:
```diff
@@ -160,22 +181,301 @@ def clean_up_cloned_repos():
+        if target:
+            logger.info("Committing image for %s" % target)
+            msg = '--message="Updated secure binary for %s (%s)"' % (target,
+                                                                    toolchain)
```
We need a space before `toolchain` to align it with `target`.

urutva replied:
Fixed

Patater commented on tools/psa/build_tfm.py:
```diff
@@ -142,15 +162,16 @@ def check_and_clone_repo(name, deps):
     check_and_clone_repo('CMSIS_5', dependencies)
-    detect_and_write_tfm_version(join(TF_M_BUILD_DIR, 'trusted-firmware-m'))
+    detect_and_write_tfm_version(join(TF_M_BUILD_DIR, 'trusted-firmware-m'),
+                                  commit)
```
We need to remove a space before `commit` to align with `join`.

urutva replied:
Fixed

Patater commented on tools/psa/build_tfm.py:
```diff
@@ -160,22 +181,301 @@ def clean_up_cloned_repos():
+                    tfm_target=None, tfm_toolchain=None, debug=False):
+
+    if configure:
+        cmake_cmd = ['cmake', '-GUnix Makefiles']
```
I'd prefer to use [Ninja](https://ninja-build.org/) for building if it can be guaranteed we have [Ninja](https://ninja-build.org/) available.

`-GNinja`

urutva replied:
@Patater I tried `-GNinja` with Cmake and generation of build files succeeds but build fails. Looks some work is required to enable Ninja in TF-M.
```[TF-M-Builder] 15:59:59: [1/217] Building C object secure_fw/CMakeFiles/tfm_s_obj_lib.dir/core/ipc/tfm_svcalls.o.
[TF-M-Builder] 15:59:59: FAILED: secure_fw/CMakeFiles/tfm_s_obj_lib.dir/core/ipc/tfm_svcalls.o .
[TF-M-Builder] 15:59:59: /Users/devran01/Documents/devaraj/compilers/gcc/gcc-arm-none-eabi-7-2018-q2-update/bin/arm-none-eabi-gcc -DCY8C624ABZI_D44=1 -DCY_PSOC6_CM0P=1 -DDAUTH_CHIP_DEFAULT -DDOMAIN_NS=__DOMAIN_NS -DNDEBUG=1 -DTFM_CORE_DEBUG -DTFM_LEGACY_API -DTFM_LVL=1 -DTFM_MULTI_CORE_TOPOLOGY=1 -DTFM_NVCOUNTERS_ENABLE -DTFM_PSA_API -D__ARM_FEATURE_CMSE=0 -D__DOMAIN_NS=0 -D__thumb2__ -I../platform/ext/cmsis -I../platform/ext/target/cypress/psoc64 -I../platform/ext/target/cypress/psoc64/CMSIS_Driver/Config -I../platform/ext/target/cypress/psoc64/Device/Config -I../platform/ext/target/cypress/psoc64/Device/Include -I../platform/ext/target/cypress/psoc64/Native_Driver/include -I../platform/ext/target/cypress/psoc64/Native_Driver/generated_source -I../platform/ext/target/cypress/psoc64/mailbox -I../platform/ext/target/cypress/psoc64/partition -I../interface/include -I../platform/include -I../secure_fw/core/arch/include -I../secure_fw/core/ipc/include -I../ -I../secure_fw/spm -I../secure_fw/core/include -I../secure_fw/core -I../secure_fw/core/ipc -I../platform/ext/common -I../platform/ext/target/cypress/psoc64/CMSIS_Driver -I../platform/ext/driver -mcpu=cortex-m0plus  -O3 -DNDEBUG   -xc -std=c99 -fshort-enums -fshort-wchar -funsigned-char -msoft-float -ffunction-sections -fdata-sections --specs=nano.specs -Wall -Werror -Wno-format -Wno-return-type -Wno-unused-but-set-variable --depend=secure_fw/CMakeFiles/tfm_s_obj_lib.dir/core/ipc/tfm_svcalls.o.d --depend_single_line --no_depend_system_headers -o secure_fw/CMakeFiles/tfm_s_obj_lib.dir/core/ipc/tfm_svcalls.o   -c ../secure_fw/core/ipc/tfm_svcalls.c.
[TF-M-Builder] 15:59:59: arm-none-eabi-gcc: error: unrecognized command line option '--depend=secure_fw/CMakeFiles/tfm_s_obj_lib.dir/core/ipc/tfm_svcalls.o.d'.
[TF-M-Builder] 15:59:59: arm-none-eabi-gcc: error: unrecognized command line option '--depend_single_line'.
[TF-M-Builder] 15:59:59: arm-none-eabi-gcc: error: unrecognized command line option '--no_depend_system_headers'; did you mean '-Wno-system-headers'?.```

Patater replied:
Saddness...

Sounds like they have bugs in their CMake files.

Patater commented on tools/psa/build_tfm.py:
```diff
@@ -160,22 +181,301 @@ def clean_up_cloned_repos():
+                        default=False)
+
+    parser.add_argument("--clean-build",
+                        help="Remove cmake build directory",
```
s/cmake/CMake

urutva replied:
Fixed

Patater commented on tools/psa/build_tfm.py:
```diff
@@ -160,22 +181,301 @@ def clean_up_cloned_repos():
-    Build Trusted Firmware M (TF-M) image for mbed-os supported TF-M targets.
-    Current version of the script only clones TF-M git repo and dependencies
-    and creates a VERSION.txt file under 'features/FEATURE_PSA/FEATURE_TFM'
+    Build Trusted Firmware M (TF-M) image for mbed-os supported TF-M targets
```
```diff
- Build Trusted Firmware M (TF-M) image for mbed-os supported TF-M targets
+ Build TrustedFirmware-M (TF-M) image for supported Mbed OS targets
```

urutva replied:
Fixed

Patater commented on tools/psa/build_tfm.py:
```diff
@@ -18,22 +18,24 @@
-
-TF_M_BUILD_DIR = join(ROOT, os.pardir, 'tfm_build_dir')
-VERSION_FILE_PATH = join(ROOT, 'features/FEATURE_PSA/FEATURE_TFM')
+TF_M_BUILD_DIR = abspath(join(ROOT, os.pardir, 'tfm_build_dir'))
```
Do we want to build into `BUILD` folder?

urutva replied:
@Patater Unfortunately it is not possible to filter out directories with Mbed OS build system. If we use `BUILD` folder then if one tries to build non-secure side (Mbed OS) after build secure side (TF-M) then Mbed OS build system will try to build secure side and fail.

Patater replied:
Please use a temp folder, via Python's tempfile API.

urutva replied:
@Patater Added new command line option `--develop`. By default a temporary folder is used as tf-m build folder. But when `--develop` command line option is used then a new folder (tfm_build_dir) under the parent folder of Mbed OS is used as tf-m build folder

Patater commented on tools/psa/build_tfm.py:
```diff
@@ -160,22 +181,301 @@ def clean_up_cloned_repos():
+            cmake_cmd.append('-DBL2=False')
+        cmake_cmd.append('..')
+    else:
+        cmake_cmd = ['cmake', '--build', '.', '--', 'install']
```
What does `install` do?

urutva replied:
@Patater Export NS APIs to a dedicated folder under cmake build folder

Patater replied:
Add a comment please.

urutva replied:
Fixed.

Patater commented on tools/targets/__init__.py:
```diff
@@ -418,8 +418,8 @@ def is_PSA_non_secure_target(self):
-    def is_PSA_target(self):
-        return self.is_PSA_secure_target or self.is_PSA_non_secure_target
+    def is_TFM_target(self):
+        return getattr(self, 'tfm_target_name', False)
```
Did single-core PSA targets use `is_PSA_target()` at all?

urutva replied:
@Patater No, it was used to identify Armv8-M and dual core PSA targets.

Patater replied:
OK

Patater commented on targets/targets.json:
```diff
@@ -9047,8 +9057,16 @@
+        "tfm_bootloader_supported": false,
+        "tfm_default_toolchain": "ARMCLANG",
+        "tfm_supported_toolchains": ["ARMCLANG", "GNUARM"],
+        "delivery_dir": "TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CPROTO_064_SB/prebuilt"
```
Why not directly into `TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CPROTO_064_SB/`?

urutva replied:
@Patater continuing the tradition of putting binaries in a dedicated folder.

Patater replied:
It's a tradition now, is it?

Mbed Bootloader doesn't do this.
`./features/FEATURE_BOOTLOADER/targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_K66F/TARGET_FRDM/mbed-bootloader-k66f-block_device-sotp-v3_4_0.bin`

Softdevice doesn't do this.
`./targets/TARGET_NORDIC/TARGET_MCU_NRF51822/device/TOOLCHAIN_IAR/s110_nrf51822_7.1.0_softdevice.bin`

Cypress WHD doesn't do this.
`./targets/TARGET_Cypress/TARGET_PSOC6/COMPONENT_WHD/resources/firmware/COMPONENT_4343W/4343WA1-mfgtest.bin`

It looks like maybe Mbed PSA used to do this, but I think it's better style not to have a `prebuilt` folder.

urutva replied:
Fixed.

Patater reviewed (changes requested):
Mostly requesting documentation changes

urutva committed:
Exporters are not supported for TF-M targets

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
Fix review comments

* Add python2 support
* Use fix version of trusted-firmware-m and it's dependencies
* Remove `--clean-build` and `-v` command-line options
* Code refactoring
* By default a temporary folder is used as tf-m build folder. But when
  `--develop` command line option is used then a new folder
  (tfm_build_dir) under the parent folder of Mbed OS is used as tf-m
  build folder
* Updated documentation
* Adding both .bin and .axf into dedicated directory for every tf-m
target

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -1,15 +1,19 @@
 # Supporting PSA in Mbed OS
 
-This document is still a work in progress. It describes the process of adding new PSA targets to Mbed OS. It also outlines the new directory structure created to support PSA on single v7-M, dual v7-M and single v8-M.
+This document is still a work in progress. It describes the process of adding new PSA targets to Mbed OS and building [TF-M](https://git.trustedfirmware.org/trusted-firmware-m.git/tree/?h=feature-twincpu) for v8-M and dual CPU targets. It also outlines the new directory structure created to support PSA on support PSA on all targets.
```
```diff
- to support PSA on support PSA on all targets.
+ to support PSA on all targets.
```

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -18,16 +22,32 @@ To help with the creation of new PSA targets, couple of generic PSA targets have
+
+An Armv8-M PSA target employs hardware to isolate the NSPE from the SPE. On Armv8-M targets, TF-M runs on the SPE and provides PSA services. Therefore, it is **MANDATORY** to define both the SPE and the NSPE target for Armv8-M PSA targets.
+
+The new PSA targets **MUST** `inherit` from the generic PSA target that corresponds to that target's architecture. Also, in case of dual v7-M and v8-M targets, SPE and NSPE targets **MUST** inherit from the respective SPE and NSPE PSA generic targets.
```
```diff
- dual v7-M
+ dual-core
```

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -40,8 +60,8 @@ Example:
-### Adding single v7-M targets
-The PSA on single v7-M is supported with the help of Mbed OS PSA services. The following is taken from `targets/targets.json` and shows a PSA enabled target, `K64F`.
+## Adding single-core PSA targets
+A target can be categorized as a single-core target if it only has a single core. Mbed OS's PSA service emulation provides PSA compatibility for single-core PSA targets. The following is taken from `targets/targets.json` and shows a PSA-enabled single-core target, `K64F`.
```
"A target can be categorized as a single-core target if it only has a single core." seems like this would be self-evident. We could remove this sentence.

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -1,33 +1,53 @@
+
+An Armv8-M PSA target employs hardware to isolate the NSPE from the SPE. On Armv8-M targets, TF-M runs on the SPE and provides PSA services. Therefore, it is **MANDATORY** to define both the SPE and the NSPE target for Armv8-M PSA targets.
+
+The new PSA targets **MUST** `inherit` from the generic PSA target that corresponds to that target's architecture. Also, in case of dual-core and v8-M targets, SPE and NSPE targets **MUST** inherit from the respective SPE and NSPE PSA generic targets.
```
Saying "the new PSA targets" begs the question: what about the old PSA targets? I would suggest instead saying what PSA targets need universally instead.

```diff
- The new PSA targets
+ PSA targets
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -101,14 +121,57 @@ The PSA on single v7-M is supported with the help of Mbed OS PSA services. The f
+
+The SPE target **MUST** contain the following attributes:
+
+* inherits: PSA generic target PSA_DUAL_V7_M_SPE
```
Use backticks to make it clear that `inherits` and `PSA_DUAL_V7_M_SPE` are literal terms that would be found in `targets.json`. Same for `tfm_target_name` and other literal terms.

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -101,14 +121,57 @@ The PSA on single v7-M is supported with the help of Mbed OS PSA services. The f
+* inherits: PSA generic target PSA_DUAL_V7_M_SPE
+* tfm_target_name: Target name in TF-M
+* tfm_bootloader_supported: If TF-M bootloader is supported by the target. Values supported are "true" and "false"
+* tfm_default_toolchain: Default TF-M toolchain supported. Values supported are "ARMCLANG" and "GNUARM"
```
Use backticks around `ARMCLANG` and `GNUARM`.

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -120,8 +183,36 @@ example mbed_app.json:
+  --clone               Clone TF-M git repo and its dependencies
+  --commit              Create a git commit for each platform
+  --clean-build         Remove Cmake build directory
+  -v, --verbose         Verbose output
```
I don't see `--develop` documented here.

urutva replied:
Fixed

Patater commented on tools/export/makefile/__init__.py:
```diff
@@ -288,18 +288,21 @@ class Armc5(Arm):
-        return arm_res or armc5_res
+        if not target.is_TFM_target:
+            if int(target.build_tools_metadata["version"]) > 0:
+                #Although toolchain name is set to ARM above we should check for ARMC5 for 5.12/onwards
```
Would recommend adding spaces after `#` in Python comments, in a separate commit from the comimt that wraps this in `if not target.is_TFM_target`

urutva replied:
Fixed

jainvikas8 commented on tools/psa/build_tfm.py:
```diff
@@ -461,21 +394,122 @@ def main():
         logger.error('"git-lfs" is not installed. Exiting...')
     elif is_cmake_installed() != 0:
-        logger.error('"cmake" is not installed. Exiting...')
+        logger.error('"Cmake" is not installed. Exiting...')
```
I think there needs to be a check for if *make* is installed or not, faced a problem on the Windows machine.

urutva replied:
Agreed. Added a check for availability of Make tool.

jainvikas8 commented on tools/psa/build_tfm.py:
```diff
@@ -160,22 +182,334 @@ def clean_up_cloned_repos():
+                         'trusted-firmware-m', 'configs/ConfigCoreIPC.cmake')))
+        cmake_cmd.append('-DTARGET_PLATFORM=' + tfm_target)
+        cmake_cmd.append('-DCOMPILER=' + tfm_toolchain)
+        if debug:
```
You can directly check the arg.debug here directly or make it global than passing it as an additional argument. This may improve code readability a bit.

urutva replied:
Fixed.

jainvikas8 commented on tools/psa/build_tfm.py:
```diff
@@ -160,22 +182,334 @@ def clean_up_cloned_repos():
+                        default=False)
+
+    parser.add_argument("--clean-build",
+                        help="Remove Cmake build directory",
```
Would be good to include this additional info "--clean-build option can only be used with --develop option"

urutva replied:
`--clean-build` option removed. Not applicable anymore.

jainvikas8 commented on tools/psa/build_tfm.py:
```diff
@@ -160,22 +182,334 @@ def clean_up_cloned_repos():
+            if TF_M_BUILD_DIR is not None:
+                clean_build_directory(TF_M_BUILD_DIR)
+            else:
+                logger.info("Build directory doesn't exists. Exiting...")
```
Do we need to exit if we don't have the TF_M_BUILD_DIR, the user could be making sure to start from scratch

urutva replied:
`--clean-build` option removed. Not applicable anymore.

jainvikas8 commented on tools/psa/build_tfm.py:
```diff
@@ -160,22 +182,334 @@ def clean_up_cloned_repos():
+    for line in iter(proc.stdout.readline, b''):
+        logger.debug(line.decode("utf-8").strip('\n'))
+
+    std_out, std_err = proc.communicate()
```
What is `std_out` used for?

urutva replied:
`proc.communicate()` returns a tuple std_out and std_err. The std_out is already consumed in the `for` loop before this hence ignored here.

urutva committed:
Update comments in python files

Add a space after `#` for comments in python files

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

jainvikas8 commented on tools/psa/build_tfm.py:
```diff
@@ -160,22 +182,334 @@ def clean_up_cloned_repos():
+                        help="Print supported TF-M secure targets")
+
+    parser.add_argument("--commit",
+                        help="Create a git commit for each platform",
```
Are we planning to support multiple commits for all supported platforms if `python3 tools/psa/build_tfm.py` used? It would be good to restrict this option only when a specific target is used?

urutva replied:
@jainvikas8 Good question.

@jainvikas8 @Patater 
While building tf-m for all targets, should we have commits for tf-m binary for every target (This will lead to `N` number of commits for every release where `N` is number of supported targets) or just one commit with tf-m binaries for all the targets included?

jainvikas8 replied:
I think the `N` number of commits should be good as it differentiates which target binaries were added. Breaking commits would be beneficial as it is easy to follow for each target.

Patater replied:
One commit is fine. The aim of the commit would be "Update TF-M to version blah blah" and would include updating all necessary headers and binaries.

jainvikas8 commented on tools/psa/build_tfm.py:
```diff
@@ -160,22 +182,334 @@ def clean_up_cloned_repos():
+        os.mkdir(cmake_build_dir)
+
+    debug = False
+    if args.mcu:
```
The if and else statement seems to have Code duplication, we could just modularise or put it in a single loop

urutva replied:
Fixed

bulislaw commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -101,14 +124,57 @@ The PSA on single v7-M is supported with the help of Mbed OS PSA services. The f
+* `inherits`: PSA generic target PSA_DUAL_V7_M_SPE
+* `tfm_target_name`: Target name in TF-M
+* `tfm_bootloader_supported`: If TF-M bootloader is supported by the target. Values supported are "true" and "false"
+* `tfm_default_toolchain`: Default TF-M toolchain supported. Values supported are `ARMCLANG` and `GNUARM`
```
Do we need to change the names of the compilers from standard Mbed ones?

urutva replied:
@bulislaw Unfortunately yes. TF-M uses different naming conventions than Mbed OS. We are using TF-M build system to generate the binary. Therefore we have to use the naming followed by TF-M. Since this is a target specific parameter, it is added to `targets.json`.

bulislaw commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -1,33 +1,56 @@
 
-## Adding new PSA targets
-To help with the creation of new PSA targets, couple of generic PSA targets have been added to `targets/targets.json`.
+# Adding new PSA targets
```
This section feels like it should be part of the porting guide in our handbook

urutva replied:
@bulislaw Yes. The idea is to capture everything about PSA support in Mbed OS in a document (supporting_psa_in_mbed-os.md) and when we are ready to merge the feature branch to master, update the porting guide as well the docs with the information already in supporting_psa_in_mbed-os.md.

urutva committed:
Change the location of tf-m build folder

Use the folder `features/FEATURE_PSA/TARGET_TFM/TARGET_IGNORE` as the
build folder to clone and build tf-m.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

jainvikas8 commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
````diff
@@ -120,8 +194,33 @@ example mbed_app.json:
+  --commit              Create a git commit for each platform
+```
+
+If the python script is invoked without any options then TF-M will be built for all the supported targets and the binary will be copied to predefined location defined by attribute `delivery_dir`.
````
It would be good to mention the `--commit`, would create a single commit for all the platforms if no targets are specified. (If we are agreeing on single commit for all the target binaries)

urutva replied:
Fixed

jainvikas8 commented on tools/psa/build_tfm.py:
```diff
@@ -18,164 +18,498 @@
+
+    :param target: Target name.
+    :param toolchain: Toolchain
+    :return: tuple (target name, tfm taret name, toolchain, delivery directory)
```
TFM target name

urutva replied:
Fixed

jainvikas8 commented on tools/psa/build_tfm.py:
```diff
@@ -18,164 +18,498 @@
+        _copy_binaries(tgt[1], cmake_build_dir, tgt[3])
+
+        if args.commit:
+            _commit_changes(tgt[3], tgt[0], tgt[2])
```
It's a bit difficult to follow what is `tgt[0], tgt[1], tgt[2].....` can we please define readable names for 0,1,2,3 - like enums or something please?

urutva replied:
Added comments to improve the code readability.

jainvikas8 commented on tools/psa/build_tfm.py:
```diff
@@ -18,164 +18,498 @@
+            logger.info("No changes detected in %s, skipping commit" %
+                                                    relpath(directory, ROOT))
+
+def _run_cmake_build(configure, cmake_build_dir, bl_supported= None,
```
bl_support=None, Please remove space.

urutva replied:
Fixed

jainvikas8 commented on tools/psa/build_tfm.py:
```diff
@@ -18,164 +18,498 @@
 
     :param command: System command as a list of tokens
+    :param output: If set to True return output from child process
+    :return: Return either outout from child process or errorcode
```
output (Typo)

urutva replied:
Fixed

jainvikas8 commented on tools/psa/build_tfm.py:
```diff
@@ -18,164 +18,498 @@
+        cmd = ['git', '-C', TF_M_BUILD_DIR, 'clone', '-b',
                deps.get(name)[1], deps.get(name)[0]]
-        _out = run_cmd_and_return_output(cmd)
+        _run_cmd_and_return(cmd)
```
Are we checking the `return code` for def - _run_cmd_and_return? If the command was successful or not, if the `output=False`? There are numerous places where we don't check.

urutva replied:
The commands are not expected to fail at places where the return value from `_run_cmd_and_return()` is not checked.

urutva committed:
Code refactoring and toolchain validation

* Refactor methods _build_tfm() and _run_cmake_build()
* Update `tfm_target_name` to psoc64_1m for CY8CPROTO_064_SB_S
* Add new attribute `"OUTPUT_EXT": "hex"` for CY8CPROTO_064_SB_S
* Update documentation

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

jainvikas8 commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -180,6 +181,8 @@ The following example shows a PSA enabled dual-core target, `PSoC64`,
 Please pay attention to the config options `extra_labels_add` and `device_has_remove`. If needed, a PSA target definition **MUST** use [extra_labels/device_has]`_add` or [extra_labels/device_has]`_remove` (not `extra_labels` or `device_has`) to add/remove either extra_labels or target capabilities. Also, use `feature_`[add/remove] to add/remove a feature.
 Check [extra_labels](https://os.mbed.com/docs/mbed-os/v5.14/reference/adding-and-configuring-targets.html), [device_has](https://os.mbed.com/docs/mbed-os/v5.14/reference/adding-and-configuring-targets.html) and [features](https://os.mbed.com/docs/mbed-os/v5.14/reference/adding-and-configuring-targets.html) for more information.
 
+By default TF-M build generates a `bin` file. If the target requires a `hex` file then the attribute `"OUTPUT_EXT": "hex"` should be added to the target definition.
```
I think we need to change this statement again, as we don't tend to add `.bin` any more and we exclusively need to define OUTPUT_EXT? Or are we giving the user the option to add hex or bin?

Patater replied:
It's target specific which format should be used for flashing to a device. We shouldn't check in all types just because (hex, bin). We should check in what will be needed. Don't check in stuff you don't need. What you need is target specific. The binaries themselves are also only useful for one target.

Patater replied:
Other targets, like v8-M ones, will most likely need either a `.a` file, or a `.bin` which we can include from an assembly file, so that we can link the SPE and NSPE images together into one single programmable binary.

urutva replied:
@jainvikas8 `I think we need to change this statement again, as we don't tend to add .bin any more`. It is not entirely true. The TF-M build process generated a `bin` file by default and we can checkin the same bin file. When generating a complete binary, we can make use of assembler instructions or post-binary hooks to combine the binary files ( bootloader if exists, secure, non-secure). Therefore this will be the default choice.
This approach is not feasible if there is a dependency on third party tools which expects a `hex` file. Adding new attribute `"OUTPUT_EXT": "hex"` is to handle those use cases.

This behaviour is same as that of Mbed OS build tools.

jainvikas8 commented on tools/psa/build_tfm.py:
```diff
@@ -18,164 +18,536 @@
 import sys
+import signal
 import shutil
+import stat
```
Are we using this? `stat`

urutva replied:
Fixed.

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -40,8 +63,16 @@ Example:
+## Naming convention for dual-core and Armv8-M target names
+As described in previous paragraphs, both SPE and NSPE target names **MUST** be defined for dual-core and Armv8-M targets. This section defines the naming convention for the same.
+
+`TargetA_S`       : PSA secure target (SPE)  
```
What is "TargetA"? Does the "A" stand for something? Will folks adding boards think they need to put an "A" after their board name? Maybe use "TargetName" instead of "TargetA" to avoid such misunderstandings..

urutva replied:
Replaced `TargetA` with `TargetName`.

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
````diff
@@ -101,14 +132,60 @@ The PSA on single v7-M is supported with the help of Mbed OS PSA services. The f
 ```
 
-Please pay attention to config options `extra_labels_add` and `device_has_add`. If needed then a PSA target definition **MUST** use [extra_labels/device_has]`_add` or [extra_labels/device_has]`_remove` (not `extra_labels` or `device_has`) to add/remove either extra_labels or target capabilities. Also, use `[feature_]`add/remove to add/remove a feature.
+Please pay attention to config options `extra_labels_add` and `device_has_add`. If needed then a PSA target definition **MUST** use [extra_labels/device_has]`_add` or [extra_labels/device_has]`_remove` (not `extra_labels` or `device_has`) to add/remove either extra_labels or target capabilities. Also, use `feature_`[add/remove] to add/remove a feature.
````
```diff
- Please pay attention to config options
+ Please pay attention to the config options
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
````diff
@@ -101,14 +132,60 @@ The PSA on single v7-M is supported with the help of Mbed OS PSA services. The f
 ```
 
-Please pay attention to config options `extra_labels_add` and `device_has_add`. If needed then a PSA target definition **MUST** use [extra_labels/device_has]`_add` or [extra_labels/device_has]`_remove` (not `extra_labels` or `device_has`) to add/remove either extra_labels or target capabilities. Also, use `[feature_]`add/remove to add/remove a feature.
+Please pay attention to config options `extra_labels_add` and `device_has_add`. If needed then a PSA target definition **MUST** use [extra_labels/device_has]`_add` or [extra_labels/device_has]`_remove` (not `extra_labels` or `device_has`) to add/remove either extra_labels or target capabilities. Also, use `feature_`[add/remove] to add/remove a feature.
````
```diff
- If needed then a PSA target
+ If needed, a PSA target
```

Patater replied:
```diff
- to add/remove either extra_labels or target capabilities
+ to add/remove either `extra_labels` or target capabilities
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
````diff
@@ -101,14 +132,60 @@ The PSA on single v7-M is supported with the help of Mbed OS PSA services. The f
+    },
+```
+
+Please pay attention to the config options `extra_labels_add` and `device_has_remove`. If needed, a PSA target definition **MUST** use [extra_labels/device_has]`_add` or [extra_labels/device_has]`_remove` (not `extra_labels` or `device_has`) to add/remove either extra_labels or target capabilities. Also, use `feature_`[add/remove] to add/remove a feature.
````
```diff
- add/remove either extra_labels or target capabilities
+ add/remove either `extra_labels` or target capabilities
```

urutva replied:
Fixed

Patater commented on targets/targets.json:
```diff
@@ -83,6 +83,11 @@
     },
     "PSA_DUAL_V7_M_SPE": {
```
Do we want to rename "PSA_DUAL_V7_M_SPE" to "PSA_DUAL_CORE_SPE" to match our documentation? Same would go for NSPE

urutva replied:
The name `PSA_DUAL_CORE_SPE` makes more sense than `PSA_DUAL_V7_M_SPE` as both cores might not be same on the targets. Updated both `targets.json` and document.

Patater commented on tools/tests.py:
```diff
@@ -24,13 +24,14 @@
-            t for t in target.supported_toolchains
-            if (t == 'ARM' or t == 'uARM')
-        ]
+    if not target.is_TFM_target:
```
Where are we documenting that tests.py can't run on TF-M targets? Do we ever want tests.py to run on TF-M targets in the future? If so, what would it take to make it work?

urutva replied:
I've created a JIRA task https://jira.arm.com/browse/IOTSEC-1265 to keep track of this requirement.

Patater commented on tools/export/cces/__init__.py:
```diff
@@ -53,9 +53,12 @@ def is_target_supported(cls, target_name):
-        return (cls.TOOLCHAIN in target.supported_toolchains) \
-            and hasattr(target, "device_name") \
-            and (target.device_name in SUPPORTED_DEVICES)
+        if not target.is_TFM_target:
```
Where are we documenting that exporters don't work with TF-M targets? Do we ever want exporters to work with TF-M targets in the future? If so, what would it take to make it work (maybe CMake)?

urutva replied:
Updated documentation to indicate that, Mbed OS exporters are currently not supported for PSA dual-core and Armv8-M targets.

 `Do we ever want exporters to work with TF-M targets in the future?`
This is a product question. If the answer is yes, I think the following could work:

* Use Mbed OS exporters to export Mbed OS related files
* Export TF-M secure side binary
* If we make assembler directives to work then linker will take care of linking both secure side and non-secure side binaries.

urutva committed:
Remove unused `import stat`

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
Update generic PSA target names

The generic PSA target names `PSA_DUAL_V7_M_SPE` and
`PSA_DUAL_V7_M_NSPE` are changed to `PSA_DUAL_CORE_SPE` and
`PSA_DUAL_CORE_NSPE` respectively.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
Update example target name in the document

Change target name `TargetA_*` to `TargetName_*`

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
Documentation fix

Add backquote to target attribute `extra_labels_add`

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>
